### PR TITLE
[dynamic_color] Add tone-based colors to color scheme

### LINKED ISSUE #574: Using `dynamic_color` on Flutter 3.21.0-1.0.pre.2 yields weird/incorrect color palette
### Package

dynamic_color

### Existing issue?

- [X] I checked the [existing issues](https://github.com/material-foundation/flutter-packages/issues)

## Expected vs actual result:
Note the elevated button, it's almost the same color as background (surface).

Screenshot_20240330-224328</img>

## Steps to reproduce:
1. Create a Flutter application project on 3.21.0-1.0.pre.2 (beta channel)
2. Add `dynamic_color: ^1.7.0` to `dependencies` in `pubspec.yaml`
3. Paste code below into `main.dart`
```dart
import 'package:dynamic_color/dynamic_color.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        Expanded(
          child: MaterialApp(
            title: 'Flutter Demo',
            theme: ThemeData(
              colorScheme: ColorScheme.fromSeed(
                seedColor: Colors.blue,
              ),
              useMaterial3: true,
            ),
            darkTheme: ThemeData(
              colorScheme: ColorScheme.fromSeed(
                seedColor: Colors.blue,
                brightness: Brightness.dark,
              ),
              useMaterial3: true,
            ),
            themeMode: ThemeMode.system,
            home: const MyHomePage(text: "ColorScheme.fromSeed"),
            debugShowCheckedModeBanner: false,
          ),
        ),
        Expanded(
          child: DynamicColorBuilder(
            builder: (light, dark) => MaterialApp(
              title: 'Flutter Demo',
              theme: ThemeData(
                colorScheme: light ?? ColorScheme.fromSeed(
                  seedColor: Colors.blue,
                ),
                useMaterial3: true,
              ),
              darkTheme: ThemeData(
                colorScheme: dark ?? ColorScheme.fromSeed(
                  seedColor: Colors.blue,
                  brightness: Brightness.dark,
                ),
                useMaterial3: true,
              ),
              themeMode: ThemeMode.system,
              home: const MyHomePage(text: "DynamicColorBuilder"),
              debugShowCheckedModeBanner: false,
            ),
          ),
        )
      ],
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.text});

  final String text;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String get text => widget.text;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(text),
            ElevatedButton(
              onPressed: () {},
              child: const Text("Click me"),
            ),
          ],
        ),
      ),
    );
  }
}
```
4. Run the app and observe the difference

### Relevant log output

```shell
[✓] Flutter (Channel beta, 3.21.0-1.0.pre.2, on Arch Linux 6.8.2-arch2-1, locale en_US.UTF-8)
    • Flutter version 3.21.0-1.0.pre.2 on channel beta at /data/distr/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision c398442c35 (3 weeks ago), 2024-03-12 22:26:24 -0700
    • Engine revision 0d4f78c952
    • Dart version 3.4.0 (build 3.4.0-190.1.beta)
    • DevTools version 2.33.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /data/distr/android-sdk
    • Platform android-VanillaIceCream, build-tools 34.0.0
    • ANDROID_HOME = /data/distr/android-sdk
    • Java binary at: /home/evgfilim1/.local/share/JetBrains/Toolbox/apps/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /usr/bin/chromium

[✓] Linux toolchain - develop for Linux desktop
    • clang version 17.0.6
    • cmake version 3.29.0
    • ninja version 1.11.1
    • pkg-config version 2.1.1

[✓] Android Studio (version 2023.3)
    • Android Studio at /home/evgfilim1/.local/share/JetBrains/Toolbox/apps/android-studio
    • Flutter plugin version 78.4.2
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.3)
    • IntelliJ at /home/evgfilim1/.local/share/JetBrains/Toolbox/apps/intellij-idea-ultimate
    • Flutter plugin version 78.4.2
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] Connected device (3 available)
    • Pixel 7 (mobile) • 192.168.0.2:34305 • android-arm64  • Android 14 (API 34)
    • Linux (desktop)  • linux             • linux-x64      • Arch Linux 6.8.2-arch2-1
    • Chrome (web)     • chrome            • web-javascript • Chromium 123.0.6312.86 Arch Linux

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

### PR DESCRIPTION
**Pull request by hasali19**
## Description

Adds the new tone-based colors from Flutter 3.22 to the color scheme.

On API >= 34, the color values are retrieved directly from the Android OS.

On API >= 31, the existing core palette will continue to be used, with the mappings taken from https://github.com/material-components/material-components-android/blob/master/docs/theming/Color.md.

## Issues
Fixes #574, #582

## Checklist

- [x] I've reviewed the [contribution guide](https://github.com/material-foundation/flutter-packages/blob/main/CONTRIBUTING.md).
- [x] I've updated the package `CHANGELOG.md`

### CODE REVIEW

**Comment by Henry-Hiles** [comment]
I'm trying to use this, but I get this error: `When a TonalPalette is created with fromList, tone must be one of [0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 95, 99, 100]`

**Comment by hasali19** [comment]
Interesting...I don't think I've changed anything related to that. Does the latest main branch work fine for you?

**Comment by Henry-Hiles** [comment]
> Interesting...I don't think I've changed anything related to that. Does the latest main branch work fine for you?

Well, from pub it doesn't have the correct colours, but no errors. I switched the pubspec to use this fork, and boom: error. I'll try main asap.

**Comment by Henry-Hiles** [comment]
I'm using DynamicColorBuilder ofc.

**Comment by hasali19** [comment]
Which flutter version are you using?

**Comment by Henry-Hiles** [comment]
```
quadradical@quadraticpc ~/D/C/qlib (master)> flutter --version
Flutter 3.22.2 • channel stable • https://github.com/flutter/flutter.git
Framework • revision nixpkgs000 () • 1970-01-01 00:00:00
Engine • revision edd8546116
Tools • Dart 3.4.3 • DevTools 2.34.3
```

**Comment by hasali19** [comment]
Ok, I think I see the problem. Support for arbitrary colour tones were added in material_color_utilities 0.11.0, but stable flutter depends on 0.8.0. I just realised I was using the beta channel, which is why I didn't run into it.

I think the solution will be to sidestep material_color_utilities and try to get the colours from the native side for the API < 34 case as well.

**Comment by Henry-Hiles** [comment]
Oh, I see, thanks!

**Comment by hasali19** [comment]
I've taken a different approach of implementing logic to "guess" tones that are missing from the common tones list, based on the implementation in material_color_utilities. This will be needed until flutter supports the new version.

@Henry-Hiles Can you confirm if this works for you now?

**Comment by Henry-Hiles** [comment]
Sure, I'll test right now!

**Comment by Henry-Hiles** [comment]
While this no longer has an error, it still has the same issues as without this PR (very similar colors so nothing looks correct)

**Comment by Henry-Hiles** [comment]
Screenshot_20240709-185210_QJobProcessor

**Comment by Henry-Hiles** [comment]
As you can see, the navbar and background are the same, which isn't normal.

**Comment by hasali19** [comment]
I can't seem to reproduce that. Do you have an example repo/code you can share?

Screenshot_1720566342

**Comment by Henry-Hiles** [comment]
> I can't seem to reproduce that. Do you have an example repo/code you can share?
> 
> Screenshot_1720566342

Hmm it happened when I upgraded to 3.22.2, I'll see if I can make a test repo.

**Review by Henry-Hiles — APPROVED** [review]

**Comment by resucutie** [comment]
After building my application using it, the primary color isn't accurate
image
image

**Comment by hasali19** [comment]
Which android version are you using?

**Comment by resucutie** [comment]
Android 14

**Comment by hasali19** [comment]
On Android 14, this should be using the primary colour that is provided directly by the system.

Could you share your flutter version, debug logs and a sample project if possible.

**Comment by JunioJsv** [comment]
There's something wrong with the implementation. I just tested it in my project, and the colors don't match those of the system, which is predominantly green. However, in the app, the predominant color is dark blue. I'm using Android 14 and Flutter 3.22.0.

WhatsApp Image 2024-07-14 at 14 37 24
WhatsApp Image 2024-07-14 at 14 37 23

**Comment by JunioJsv** [comment]
Here's the link to the project where I ran the test: https://github.com/JunioJsv/minimum-flutter

image

**Comment by hasali19** [comment]
I'm not able to reproduce the issue on my device.

Could you share the debug logs when you run the app. In particular, do you see a line similar to:

```
I/flutter (18079): dynamic_color: Color scheme detected.
```

**Comment by hasali19** [comment]
It seems this may be a common issue on Samsung devices, affecting the official Android material library as well - https://github.com/material-components/material-components-android/issues/3924. Samsung reports incorrect values for the new colours added in Android 14.

Do you see dynamic colours working correctly in other apps, e.g. Google apps?

Given this is a Samsung specific issue, not really sure what we can do here.

**Comment by JunioJsv** [comment]
> It seems this may be a common issue on Samsung devices, affecting the official Android material library as well - [material-components/material-components-android#3924](https://github.com/material-components/material-components-android/issues/3924). Samsung reports incorrect values for the new colours added in Android 14.
> 
> Do you see dynamic colours working correctly in other apps, e.g. Google apps?
> 
> Given this is a Samsung specific issue, not really sure what we can do here.

Indeed, the Google apps are also displaying the wrong tone, dark blue, and my device is a Samsung A54.

**Comment by resucutie** [comment]
I've decided to finally publish a "just build it" branch of my project: https://github.com/resucutie/localbooru/tree/dynamic-colors-pr-fix-test

@hasali19 feel free to build and test it

**Comment by hasali19** [comment]
@resucutie Thanks, I'll take a look.

Could you also share your flutter version please?

**Comment by hasali19** [comment]
From what I can see this is working as expected; Android's quick settings tiles use a slightly different shade of the primary colour.

You can see [here](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/packages/SystemUI/res/values/styles.xml;l=432) that the `shadeActive` colour (used by the Android system for the quick settings tiles) uses tone 90 of the primary colour, while the [material library](https://github.com/material-components/material-components-android/blob/master/docs/theming/Color.md) defines primary in the colour scheme as tone 40 in light mode and 80 in dark mode.

**Comment by resucutie** [comment]
Welp, what about in comparasion to other apps? It looks odd, or I guess it is the consensus here, like if it is using the wrong color shade, but I am not sure

**Comment by hasali19** [comment]
Do you mean you are seeing different colours compared to other apps using the same dynamic colour scheme?

This implementation should be using the same colours as those used by the official native material library. If this is not the case, please share a repro.

**Comment by resucutie** [comment]
I've (finally) tested it and apparently it indeed take the correct colors. It is a bit odd that it looks different on an emulator




PR is good I think

**Comment by wildsylvan** [comment]
in our case, after activating dynamic colors:
image
primary container e secondary container became the same color!

---

look at the cards (list a/b and the new list fab. they should be different colors but with dynamic colors they become the same)

### without dynamic colors (correct)
Screenshot_2024-08-09-07-07-49-47_304e098f875053d2a3a592b681756698

### with dynamic colors (wrong)
Screenshot_2024-08-09-07-07-28-30_304e098f875053d2a3a592b681756698

**Comment by hasali19** [comment]
Please can you share your flutter and android versions, device information, debug logs and a link to a code repo if possible.

**Comment by wildsylvan** [comment]
thanks for the quick reply! here:

`flutter doctor`
```
[√] Flutter (Channel beta, 3.24.0-0.2.pre, on Microsoft Windows [Version 10.0.22631.3880], locale en-US)
[√] Windows Version (Installed version of Windows is version 10 or higher)
[√] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
[√] Chrome - develop for the web
[X] Visual Studio - develop Windows apps
    X Visual Studio not installed; this is necessary to develop Windows apps.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components
[√] Android Studio (version 2024.1)
[√] VS Code (version 1.92.1)
[√] Connected device (3 available)
[√] Network resources
```

testing on a real phone, `oneplus 8t` with `android 14`
image

---

i can provide a repro next week... it's very vanilla use of material components tho, not many customizations

**Comment by wildsylvan** [comment]
this is the oneplus color settings:
image

**Comment by wildsylvan** [comment]
ok these impact how colors are shown in the app:
image

but the difference between primary container e secondary container remains basically non-existent

**Comment by wildsylvan** [comment]
@hasali19 what would be the best way to ignore completely oneplus color presets, just get the "main accent color" from android and utilize that color to build a correct material 3 color palette?

**Comment by hasali19** [comment]
Thanks for the info. Given that you are using Android 14 where the colours are retrieved directly from the system without any conversion, it is likely OnePlus is simply reporting both colours as the same.

One thing you could try is to force use of the fallback method used for older versions of Android, e.g:
```dart
final corePalette = await DynamicColorPlugin.getCorePalette();
final lightDynamic = corePalette?.toColorScheme(brightness: Brightness.light);
final darkDynamic = corePalette?.toColorScheme(brightness: Brightness.dark);
```

If you just want a specific "accent" colour, you can pick one from the fields on the colour schemes provided by `DynamicColorBuilder`.

**Comment by wildsylvan** [comment]
_update_

---

```dart
final corePalette = await DynamicColorPlugin.getCorePalette();
final lightDynamic = corePalette?.toColorScheme(brightness: Brightness.light);
final darkDynamic = corePalette?.toColorScheme(brightness: Brightness.dark);
```
doesn't fix the issue

---

my colleague tested with his Pixel 7 Pro (and also inside an emulator) and he has the same issue

the issue seems to be present **only in light mode** tho (and our theme settings don't have any specific setting for either of them):

# wrong in light mode
image

# correct in dark mode
image

**Comment by wildsylvan** [comment]
this is the card (the one with the text Test):
image

and this is the fab:
image

**Comment by hasali19** [comment]
Although it's subtle, the colours in your screenshot do appear to be slightly different. I'm also able to reproduce very similar looking colours in light mode when running in an emulator.

image

**Comment by wildsylvan** [comment]
yeah i was just about to write it! they are a little different... so it's an more general issue with the way the m3 color generator work?

**Comment by wildsylvan** [comment]
with a green wallpaper the difference is less subtle too...

image

**Comment by hasali19** [comment]
Right, I guess it's just an artifact of whatever algorithm Android uses to pick the system colours.

**Comment by wildsylvan** [comment]
well to be fair the material theme builder shows it too with the blues (tho they difference is a little more visible):
image

while the green is much less subtle:
image

@guidezpl is it supposed to be that different between hues?

**Comment by im-trisha** [comment]
It seems like the card color is incorrect... Is it just me or...?

Not just the card color, but also the bottom nav bar and many other colors seem to be washed off (see, the one on the left is way more purple)

image

**Comment by hasali19** [comment]
Given that nobody appears to be maintaining this repo, I have published my fork to pub.dev as [dynamic_system_colors](https://pub.dev/packages/dynamic_system_colors).

**Comment by guidezpl** [comment]
Hi, this repo is maintained sporadically when I have time, but also it is summer and I am on vacation :) please be patient and I will review when back

File: packages/dynamic_color/lib/src/corepalette_to_colorscheme.dart [hunk 1]
```diff
@@ -9,47 +6,137 @@ extension CorePaletteToColorScheme on CorePalette {
...
-        break;
+        return _toDarkColorScheme(this);
+    }
+  }
+}
+
+ColorScheme _toLightColorScheme(CorePalette corePalette) {
+  return ColorScheme(
+    brightness: Brightness.light,
+    primary: Color(corePalette.primary.get(40)),
```
Comment (guidezpl) [inline packages/dynamic_color/lib/src/corepalette_to_colorscheme.dart:21]: Unfortunately, these values only work for one type of scheme, there are many types. See https://github.com/material-foundation/material-color-utilities/blob/main/dev_guide/creating_color_scheme.md

File: packages/dynamic_color/lib/src/corepalette_to_colorscheme.dart [hunk 2]
```diff
@@ -9,47 +6,137 @@ extension CorePaletteToColorScheme on CorePalette {
...
+    inverseSurface: Color(corePalette.neutral.get(90)),
+    onInverseSurface: Color(corePalette.neutral.get(20)),
+    inversePrimary: Color(corePalette.primary.get(40)),
+    outline: Color(corePalette.neutralVariant.get(60)),
+    outlineVariant: Color(corePalette.neutralVariant.get(30)),
+  );
+}
+
+// This logic is taken from material_color_utilities 0.12 - https://github.com/material-foundation/material-color-utilities/blob/be615fc90286787bbe0c04ef58a6987e0e8fdc29/dart/lib/palettes/tonal_palette.dart#L93C5-L111.
+// Once flutter updates to the latest version, this workaround can be removed.
```
Comment (guidezpl) [inline packages/dynamic_color/lib/src/corepalette_to_colorscheme.dart:117]: Best to wait for https://github.com/flutter/flutter/pull/170000 rather than introduce workarounds IMO

File: packages/dynamic_color/lib/src/color_schemes.dart [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  const ColorSchemes({required this.light, required this.dark});
+
+  factory ColorSchemes.fromList(List<int> colors) {
+    return ColorSchemes(
+      light: _colorSchemeFromList(colors, 0, Brightness.light),
+      dark: _colorSchemeFromList(colors, 1, Brightness.dark),
+    );
+  }
+}
+
```
Comment (guidezpl) [inline packages/dynamic_color/lib/src/color_schemes.dart:16]: This is somewhat fragile an difficult to validate correctness with DynamicColorPlugin.kt

File: packages/dynamic_color/android/src/main/kotlin/io/material/plugins/dynamic_color/DynamicColorPlugin.kt [hunk 4]
```diff
@@ -120,4 +129,98 @@ class DynamicColorPlugin : FlutterPlugin, MethodCallHandler {
       resources.getColor(android.R.color.system_neutral2_0, null),
     );
   }
+
+  @RequiresApi(Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
+  private fun getColorScheme(resources: Resources): IntArray {
+    return intArrayOf(
+      // light
```
Comment (guidezpl) [inline packages/dynamic_color/android/src/main/kotlin/io/material/plugins/dynamic_color/DynamicColorPlugin.kt:136]: Why not obtain `system_palette_key_color_primary_light` and other key colors and generate the scheme directly?

**Review by guidezpl — CHANGES_REQUESTED** [review]

**Comment by guidezpl** [comment]
To anyone on this PR with issues, please create a new separate issue if appropriate, or refer to an existing one, it's really hard to follow the discussion here

File: packages/dynamic_color/lib/src/corepalette_to_colorscheme.dart (hunk 1 above)
Comment (hasali19) [reply packages/dynamic_color/lib/src/corepalette_to_colorscheme.dart:21]: I've read through that page and I confess I'm not entirely sure what you mean. The point of this code is to generate a _specific_ colour scheme based on the system colour palette. The values used here are taken from the official Android material library - https://github.com/material-components/material-components-android/blob/master/docs/theming/Color.md. The dynamic colour implementation in Jetpack Compose is also [implemented](https://github.com/androidx/androidx/blob/androidx-main/compose/material3/material3/src/androidMain/kotlin/androidx/compose/material3/DynamicTonalPalette.android.kt#L262-L305) in a similar way.

File: packages/dynamic_color/android/src/main/kotlin/io/material/plugins/dynamic_color/DynamicColorPlugin.kt (hunk 4 above)
Comment (hasali19) [reply packages/dynamic_color/android/src/main/kotlin/io/material/plugins/dynamic_color/DynamicColorPlugin.kt:136]: Is that guaranteed to return the same colours as this code that is querying from the system? Again, this code is based on the official Android implementation [here](https://github.com/androidx/androidx/blob/androidx-main/compose/material3/material3/src/androidMain/kotlin/androidx/compose/material3/DynamicTonalPalette.android.kt#L311-L409), and I think it would be good to keep the implementation as similar as possible to ensure consistency with non-Flutter apps.

File: packages/dynamic_color/lib/src/color_schemes.dart (hunk 3 above)
Comment (hasali19) [reply packages/dynamic_color/lib/src/color_schemes.dart:16]: Good point, I'll see if I can improve that. Perhaps a map with readable names instead of a list?

**Comment by hasali19** [comment]
Thanks for the review @guidezpl! I've responded to your comments - main theme is that I've tried to keep the implementation close to what the official Android dynamic colour implementation does, to ensure consistency with non-flutter apps.

**Comment by guidezpl** [comment]
With recent changes and 1.9.0 out, I'm going to close this. Thanks!
